velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -132,12 +132,16 @@ ::duckdb::Value duckValueAt<TypeKind::HUGEINT>(
     const VectorPtr& vector,
     vector_size_t index) {
   using T = typename KindToFlatVector<TypeKind::HUGEINT>::WrapperType;
-  auto type = vector->type()->asLongDecimal();
   auto val = vector->as<SimpleVector<T>>()->valueAt(index);
   auto duckVal = ::duckdb::hugeint_t();
   duckVal.lower = (val << 64) >> 64;
   duckVal.upper = (val >> 64);
-  return ::duckdb::Value::DECIMAL(duckVal, type.precision(), type.scale());
+  if (vector->type()->isLongDecimal()) {
+    auto type = vector->type()->asLongDecimal();
+    return ::duckdb::Value::DECIMAL(duckVal, type.precision(), type.scale());
+  }
+  // Flat vector is HUGEINT type and not the logical decimal type.
+  return ::duckdb::Value::HUGEINT(duckVal);
```

Comment:
Yes, it does. This change is required to materialize duckdb results back to velox results during comparison.

Reply:
Hi @karteekmurthys, I only found that duckValueAt() is used in DuckDbQueryRunner::createTable() that create a DuckDB table for given Velox vectors. Where do you see this method is used when materializing DuckDB results back to Velox?

Reply:
Here is a stack trace that I got while running MinMaxBy tests:

<img width="955" alt="Screenshot 2024-04-30 at 9 28 39 PM" src="https://github.com/facebookincubator/velox/assets/6971561/20a18fe3-1bf1-46eb-a0da-7ae609036c5d">

Reply:
Hi @karteekmurthys, thank you for sharing the stack trace. I saw that variantAt<>() is called in this stack but not duckValueAt<>(). Do I miss anything?

Reply:
I have only encountered in the case of materializing duckdb results at the time of comparison. Not sure about duckValueAt<>() case. Sorry about the confusion.

Reply:
I found the call stack of this duckValueAt<>() method. It is called when creating a DuckDB table for the given input of RowVectors. In MinMaxAggregationTest.cpp we create FlatVector of HugeintType in the input, so `auto type = vector->type()->asLongDecimal()` doesn't work in this case. Would it make sense to convert `type` to just Decimal, e.g., `auto type = vector->type()->asDecimal()`? Then you can pass `DECIMAL(38, x)` to makeFlatVector() in the buildDataVector() method to create Decimal-typed FlatVectors.

Reply:
@kagamiori We used HUGEINT/BIGINT type as min_by and max_by just works on the Physical type. For these functions, it doesn't matter if we create Decimal or HUGEINT type vectors. 

About your suggestion, we should rather fix the way value is read for HUGEINT instead of switching to Decimals. Lmk what you think.

Reply:
Hi @karteekmurthys, in addition to the question above, I'm also curious about the Decimal type. I saw that we have DECIMAL(uint8_t precision, uint8_t scale) for creating a Decimal TypePtr, and we have Type::asShortDecimal() and Type::asLongDecimal() for converting a Type to a ShortDecimalType& or LongDecimalType&. But we don't have LONG_DECIMAL(precision, scale), SHORT_DECIMAL(precision, scale), or Type::asDecimal(). Is this by design that we should never create types of LongDecimal and ShortDecimal directly and should never use a Decimal type directly? Or we just didn't add these utility methods because we haven't needed them?

Reply:
We could try that but the rowType_ has HUGEINT in it.
We create DuckDB tables from the `rowVectors_`. With existing way of handling duckValueAt<int128_t> we always expect the type to be a long decimal. That is why we need this change to support HUGEINTs.
```

void MinMaxByAggregationTestBase::SetUp() {
 rowVectors_ = makeVectors(rowType_, 5, 10);
  createDuckDbTable(rowVectors_);
  }
```
<img width="1067" alt="Screenshot 2024-05-07 at 1 34 20 PM" src="https://github.com/facebookincubator/velox/assets/6971561/cca451d9-2fe2-4269-a4c1-3fe7f06a8fb1">

Reply:
> Hi @karteekmurthys, in addition to the question above, I'm also curious about the Decimal type. I saw that we have DECIMAL(uint8_t precision, uint8_t scale) for creating a Decimal TypePtr, and we have Type::asShortDecimal() and Type::asLongDecimal() for converting a Type to a ShortDecimalType& or LongDecimalType&. But we don't have LONG_DECIMAL(precision, scale), SHORT_DECIMAL(precision, scale), or Type::asDecimal(). Is this by design that we should never create types of LongDecimal and ShortDecimal directly and should never use a Decimal type directly? Or we just didn't add these utility methods because we haven't needed them?

That is a good question. It is by design AFAIK. In older versions we did have interface to create Short and Long Decimal types. Later we abstracted with a single DECIMAL() interface. 

`ShortDecimalType/LongDecimalType` are logical types. We wanted to distinguish between logical and physical types.

Refer: https://github.com/prestodb/presto/pull/19504

Reply:
> We could try that but the rowType_ has HUGEINT in it. We create DuckDB tables from the `rowVectors_`. With existing way of handling duckValueAt<int128_t> we always expect the type to be a long decimal. That is why we need this change to support HUGEINTs.
> 
> ```
> 
> void MinMaxByAggregationTestBase::SetUp() {
>  rowVectors_ = makeVectors(rowType_, 5, 10);
>   createDuckDbTable(rowVectors_);
>   }
> ```

It's because `createScalarType(TypeKind::HUGEINT)` returns HugeintType. I think it would be better to create Decimal-typed vectors instead of HugeInt-typed vectors since we're testing for decimal. Could you add a utility method like the following and replace every createScalarType(kind) call with createType(kind) in MinMaxByAggregationTest.cpp? 
```
TypePtr createType(TypeKind kind) {
  if (kind == TypeKind::HUGEINT) {
    return DECIMAL(38, 10);
  } else {
    return createScalarType(kind);
  }
}
```

Also, give this DECIMAL(38, 10) to makeFlatVector in buildDataVector<int128_t>() as well, e.g., `makeFlatVector<int128_t>(size, [&](auto row) { return HugeInt::build(values[row], values[row]);}, nullptr, DECIMAL(38, 0))`.

> Would it make sense to convert type to just Decimal, e.g., auto type = vector->type()->asDecimal()?

No need for this change since DECIMAL(38, 10) already creates a LongDecimalType. Once you make the change above, the original code in duckValueAt<TypeKind::HUGEINT>() before this PR would work.

Reply:
> Hi @karteekmurthys, It's because `createScalarType(TypeKind::HUGEINT)` returns HugeintType. I think it would be better to create Decimal-typed vectors instead of HugeInt-typed vectors since we're testing for decimal. Could you add a utility method like the following and replace every createScalarType(kind) call with createType(kind) in MinMaxByAggregationTest.cpp?

We are intentionally not testing the logical types. We don't need logical type info to test min_by/max_by functions. We made changes here: https://github.com/facebookincubator/velox/pull/8999 specifically to make this test Physical type only.

Reply:
@karteekmurthys, I see. Thank you for sharing this context.